apps/mark-scan/backend/src/custom-paper-handler/state_machine.ts

```diff
@@ -0,0 +1,266 @@
+import makeDebug from 'debug';
+import { assert, Optional } from '@votingworks/basics';
+import { Buffer } from 'buffer';
+import { pdfToImages, writeImageData } from '@votingworks/image-utils';
+import { createImageData } from 'canvas';
+import {
+  getPaperHandlerDriver,
+  PaperHandlerDriver,
+  PaperHandlerStatus,
+  chunkBinaryBitmap,
+  ImageConversionOptions,
+  imageDataToBinaryBitmap,
+  Uint8,
+} from '@votingworks/custom-paper-handler';
+import { SimpleServerStatus, SimpleStatus } from './types';
+
+const serverDebug = makeDebug('custom-paper-handler:machine');
+
+function debug(message: string) {
+  serverDebug(message);
+  console.log(message);
+}
+
+function isPaperInInput(paperHandlerStatus: PaperHandlerStatus): boolean {
+  return (
+    paperHandlerStatus.paperInputLeftInnerSensor ||
+    paperHandlerStatus.paperInputLeftOuterSensor ||
+    paperHandlerStatus.paperInputRightInnerSensor ||
+    paperHandlerStatus.paperInputRightOuterSensor
+  );
+}
+
+function isPaperReadyToLoad(paperHandlerStatus: PaperHandlerStatus): boolean {
+  return (
+    paperHandlerStatus.paperInputLeftInnerSensor &&
+    paperHandlerStatus.paperInputLeftOuterSensor &&
+    paperHandlerStatus.paperInputRightInnerSensor &&
+    paperHandlerStatus.paperInputRightOuterSensor
+  );
+}
+
+function isPaperInScanner(paperHandlerStatus: PaperHandlerStatus): boolean {
+  return (
+    paperHandlerStatus.paperPreCisSensor ||
+    paperHandlerStatus.paperPostCisSensor ||
+    paperHandlerStatus.preHeadSensor ||
+    paperHandlerStatus.paperOutSensor ||
+    paperHandlerStatus.parkSensor ||
+    paperHandlerStatus.paperJam ||
+    paperHandlerStatus.scanInProgress
+  );
+}
+
+function isPaperAnywhere(paperHandlerStatus: PaperHandlerStatus): boolean {
+  return (
+    isPaperInInput(paperHandlerStatus) || isPaperInScanner(paperHandlerStatus)
+  );
+}
+
+export class PaperHandlerStateMachine {
```

Comment:
That's a good point. We lock at the web device level in the driver to prevent simultaneous data transfer but nothing prevents callers from simultaneously invoking 2 commands that use different channels, or even 2 commands that use the same channel but happen to release the lock before completion. I'll think about how to do this.

Reply:
Recalling our previous conversation about updating the driver channel locks, this is a good opportunity to
* factor out `libs/custom-scanner/mutex`
* update the paper handler driver to use `Mutex` for its channels
* update the mark-scan state machine to use `Mutex` for a single global lock that controls access to any driver function

Reply:
@eventualbuddha actually, since I'm in the process of rewriting the state machine with `xstate` it might make sense to do the above in a separate PR. Thoughts?

Reply:
I'm okay with either as long as it doesn't get lost in the shuffle.

Reply:
tracking in https://github.com/votingworks/vxsuite/issues/3703